R/packages.R

```diff
@@ -0,0 +1,282 @@
+      url = library_metadata[[1]]$browser_download_url
+    )
+  )
+}
+
+# Lookup URL and metadata for Wasm binary package
+prepare_wasm_metadata <- function(pkg, metadata, verbose) {
+  desc <- utils::packageDescription(pkg)
+  repo <- desc$Repository
+  prev_ref <- metadata$ref
+  prev_cached <- metadata$cached
+  metadata$name <- pkg
+  metadata$version <- desc$Version
+
+  # Skip base R packages
+  if (!is.null(desc$Priority) && desc$Priority == "base") {
+    metadata$ref <- glue::glue("{metadata$name}@{metadata$version}")
+    metadata$type <- "base"
+    metadata$cached <- prev_cached <- TRUE
+    if (verbose) {
+      message("Skipping base R package: ", metadata$ref)
+    }
+    return(metadata)
+  }
+
+  # Set a package ref for caching
+  if (!is.null(desc$RemoteType) && desc$RemoteType == "github") {
+    user <- desc$RemoteUsername
+    repo <- desc$RemoteRepo
+    sha <- desc$RemoteSha
+    metadata$ref <- glue::glue("github::{user}/{repo}@{sha}")
+  } else if (repo == "CRAN") {
+    metadata$ref <- glue::glue("{metadata$name}@{metadata$version}")
+  } else if (grepl("Bioconductor", repo)) {
+    metadata$ref <- glue::glue("bioc::{metadata$name}@{metadata$version}")
+  } else if (grepl("r-universe\\.dev$", repo)) {
+    metadata$ref <- glue::glue("{repo}::{metadata$name}@{desc$RemoteSha}")
+  } else {
+    metadata$ref <- glue::glue("{metadata$name}@{metadata$version}")
+  }
+
+  # If not cached, discover Wasm binary URLs
+  if (is.null(prev_cached) || !prev_cached || prev_ref != metadata$ref) {
+    metadata$cached <- FALSE
+    if (!is.null(desc$RemoteType) && desc$RemoteType == "github") {
+      metadata$assets <- get_github_wasm_assets(desc)
+      metadata$type <- "library"
+    } else if (grepl("r-universe\\.dev$", repo)) {
+      metadata$assets <- get_r_universe_wasm_assets(desc)
+      metadata$type <- "package"
+    } else {
+      # Fallback to repo.r-wasm.org lookup for CRAN and anything else
+      metadata$assets <- get_default_wasm_assets(desc)
+      metadata$type <- "package"
+    }
+  } else if (verbose) {
+    message("Skipping cached Wasm binary: ", metadata$ref)
+  }
+
+  metadata
+}
+
+download_wasm_packages <- function(appdir, destdir, verbose, package_cache) {
+  verbose_print <- if (verbose) message else list
+  # App dependencies, ignoring shiny packages in base webR image
+  pkgs <- unique(renv::dependencies(appdir, quiet = !verbose)$Package)
+  pkgs <- pkgs[pkgs != "shiny" & pkgs != "bslib"]
+  if (length(pkgs) > 0) {
+    pkgs <- resolve_dependencies(pkgs, verbose)
+  }
+
+  if (verbose) {
+    p <- progress::progress_bar$new(
+      format = "[:bar] :percent\n",
+      total = length(pkgs),
+      clear = TRUE
+    )
+  }
+
+  # Create empty R packages directory in app assets if not already there
+  pkg_dir <- fs::path(destdir, "shinylive", "webr", "packages")
+  if (!fs::dir_exists(pkg_dir)) {
+    verbose_print("Creating ", pkg_dir, "/")
+    fs::dir_create(pkg_dir)
+  }
```

Comment:
Always attempt make the directory as the location is printed in the next line

```suggestion
  fs::dir_create(pkg_dir, recurse = TRUE)
```